README
RPC::ExtDirect::Client
======================

    This module implements Ext.Direct client in pure Perl. Its main purpose
    is to be used for testing server side Ext.Direct classes, complementing
    RPC::ExtDirect::Server. It can also be used for testing Ext.Direct
    Actions developed in languages other than Perl, or for something else
    altogether.

    This client is synchronous, i.e. all method calls are blocking. If you
    need a non-blocking way of calling an Ext.Direct server, take a look at
    RPC::ExtDirect::Client::Async instead.

    Unless configured otherwise, RPC::ExtDirect::Client will use HTTP::Tiny
    for the transport layer.

INSTALLATION

    To install this module type the following:

        perl Makefile.PL
        make && make test
        make install

COPYRIGHT AND LICENSE

    Copyright (c) 2012-2015 Alex Tokarev <tokarev@cpan.org>.

    This module is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself. See perlartistic.



Changes
1.24  Wed Oct 12 22:02:14 PDT 2016
    - Added a client option to specify connection protocol other than HTTP
    - Connection timeout in unit tests lowered to 1 second
    - Fixed tests to use ref_ok instead of isa_ok
    - Refactored Client.pm to avoid using `each` hash iterator
    - Version bumped to 1.24 to align with Server and core modules

1.20  Thu May 14 22:10:34 PDT 2015
    - Updated to support metadata feature in RPC::ExtDirect 3.20+
    - More tests, as usual

1.03  Tue Jun 24 21:23:54 2014
    - Client initialization with a local API instance works properly now,
      with a bunch of tests added

1.02  Mon Jun 23 17:49:55 2014
    - Fixed a typo in Makefile.PL that caused build failure when
      ExtUtils::MakeMaker version happened to be between 6.46 and 6.64

1.01  Fri Jun 20 21:18:36 2014
    - Ext.Direct Method argument checking now happens in the Method
      itself; Client is updated to take advantage of that
    - Minimal RPC::ExtDirect version bumped to 3.01 for above changes
    - Improved tests, added many new ones
    - Miscellaneous fixes

1.00  Thu Jun 12 21:47:21 2014
    - Updated to accommodate for the changes in RPC::ExtDirect 3.0
      and RPC::ExtDirect::Server 1.0
    - Major internal refactoring and tune-up to support the async
      client as a subclass (See RPC::ExtDirect::Client::Async)
    - Many, many improvements and bug fixes
    - Many new tests
    - Tons of documentation
    - Compatible with Perl 5.6.2.

0.22  Fri Mar 29 18:35:48 2013
    - Added handling for Methods that do not specify how they need
      their parameters. We assume by-name without strict checking.

0.21  Wed Mar 27 22:15:03 2013
    - Published get_api method. Made all code and tests runnable
      under Perl 5.8.

0.2   Fri Nov  9 23:52:39 2012
    - Added an option to pass cookies either with HTTP::Cookies object
      or raw hashref

0.01  Wed Jun 20 18:13:21 2012
	- original version; code runs and tests pass but lacks documentation



t/server.pl
#!/usr/bin/perl

# This script can be used for manual HTTP server testing in case
# something goes awry

use common::sense;

use RPC::ExtDirect::Server;

my $server = RPC::ExtDirect::Server->new(static_dir => 't/htdocs');
my $port   = $server->port;

say "Listening on port $port";

$server->run();



t/lib/util.pm
package util;

use Carp;

use RPC::ExtDirect::Server;

use base 'Exporter';

our @EXPORT = qw(
    start_server
    stop_server
);

our ($SERVER_PID, $SERVER_PORT);

sub start_server {
    my (%params) = @_;

    return $SERVER_PORT if $SERVER_PORT;

    my $server = RPC::ExtDirect::Server->new(%params);
    my $port   = $SERVER_PORT = $server->port;

    if ( my $pid = $SERVER_PID = fork ) {
        local $SIG{CHLD} = sub { waitpid $pid, 0 };

        # Give the child some head start
        select undef, undef, undef, 0.1;

        return $port;
    }
    elsif ( defined $pid && $pid == 0 ) {

        # Trap last breaths to avoid cluttering the screen
        local $SIG{__DIE__} = sub {};

        $server->run();

        exit 0;
    }
    else {
        croak "Can't fork: $!";
    };

    return;
}

sub stop_server {
    my ($pid) = @_;

    $pid = $SERVER_PID unless defined $pid;

    # This is a bit cruel but somehow if I use any other signal
    # the server kid has last opportunity to cry for help.
    # Which we don't want because it breaks TAP output.
    kill 9, $pid if defined $pid;

    $SERVER_PID = $SERVER_PORT = undef;
}

END { stop_server }

1;




lib/RPC/ExtDirect/Client/API.pm
package RPC::ExtDirect::Client::API;

use strict;
use warnings;
no  warnings 'uninitialized';   ## no critic

use JSON;

use RPC::ExtDirect::Util::Accessor;
use RPC::ExtDirect::Config;
use RPC::ExtDirect::API;

use base 'RPC::ExtDirect::API';

### PUBLIC CLASS METHOD (CONSTRUCTOR) ###
#
# Instantiate a new API declaration from JavaScript code
#

sub new_from_js {
    my ($class, %params) = @_;
    
    my $config = delete $params{config} || RPC::ExtDirect::Config->new();
    my $js     = delete $params{js};
    
    my $api_json = _strip_js($js, $config->remoting_var);
    my $api_href = _decode_api($api_json);
    
    my $self = $class->SUPER::new_from_hashref(
        config   => $config,
        api_href => $api_href->{actions},
        type     => $api_href->{type},
        url      => $api_href->{url},
        %params,
    );
    
    return $self;
}

### PUBLIC INSTANCE METHOD ###
#
# Return an Action object or a list of Action objects, depending on context
#

sub actions {
    my ($self, @actions) = @_;

    if ( wantarray ) {
        my @set = @actions ? @actions
                :            keys %{ $self->{actions} }
                ;
        
        my @result = map { $self->get_action_by_name($_) } @set;

        return @result;
    }
    else {
        my $action_name = shift @actions;
        
        return $self->get_action_by_name($action_name);
    };
}

### PUBLIC INSTANCE METHODS ###
#
# Read-write accessors
#

RPC::ExtDirect::Util::Accessor->mk_accessors(
    simple => [qw/ type url /],
);

############## PRIVATE METHODS BELOW ##############

### PRIVATE PACKAGE SUBROUTINE ###
#
# Extracts actual JSONified API declaration from JavaScript string
#

sub _strip_js {
    my ($js, $remoting_var) = @_;

    # We assume that the API definition is simple JavaScript
    my @parts = split /;\s*/, $js;
    
    my ($api_def) = grep /^\s*$remoting_var\s*=\s*{/, @parts;
    
    die "Can't find the API definition for $remoting_var\n"
        unless defined $api_def;
    
    # This should leave only the API object, more or less
    $api_def =~ s/^\s*$remoting_var\s*=\s*//;

    return $api_def;
}

### PRIVATE PACKAGE SUBROUTINE ###
#
# Decode API declaration and check basic constraints
#

sub _decode_api {
    my ($js) = @_;

    my ($api_js) = eval { JSON->new->utf8(1)->decode_prefix($js) };

    die "Can't decode API declaration: $@\n" if $@;

    die "Empty API declaration\n"
        unless 'HASH' eq ref $api_js;

    die "Unsupported API type\n"
        unless $api_js->{type} =~ /remoting/i;
    
    # Convert the JavaScript API definition to the format
    # API::new_from_hashref expects
    my $actions = $api_js->{actions};
    
    my %remote_actions
        = map { $_ => _convert_action($actions->{$_}) } keys %$actions;
    
    $api_js->{actions} = \%remote_actions;
    
    return $api_js;
}

### PRIVATE PACKAGE SUBROUTINE ###
#
# Convert JavaScript Action definition to remote Action definition
#

sub _convert_action {
    my ($action_def) = @_;
    
    my %methods = map { delete $_->{name} => $_ } @$action_def;
    
    return {
        remote  => 1,
        methods => \%methods,
    };
}

1;


lib/RPC/ExtDirect/Client/Async.pm
package RPC::ExtDirect::Client::Async;

use strict;
use warnings;
no  warnings 'uninitialized';

use Carp;
use File::Spec;
use HTTP::Tiny;
use AnyEvent::HTTP;

use RPC::ExtDirect::Util::Accessor;
use RPC::ExtDirect::Config;
use RPC::ExtDirect::API;
use RPC::ExtDirect;
use RPC::ExtDirect::Client;

use base 'RPC::ExtDirect::Client';

#
# This module is not compatible with RPC::ExtDirect < 3.0
#

croak __PACKAGE__." requires RPC::ExtDirect 3.0+"
    if $RPC::ExtDirect::VERSION < 3.0;

### PACKAGE GLOBAL VARIABLE ###
#
# Module version
#

our $VERSION = '3.00_01';

### PUBLIC CLASS METHOD (CONSTRUCTOR) ###
#
# Instantiate a new Async client, connect to the specified server
# and initialize the Ext.Direct API. Optionally fire a callback
# when that's done
#

sub new {
    my ($class, %params) = @_;
    
    my $api_cb = delete $params{api_cb};
    
    my $self = $class->SUPER::new(%params);
    
    $self->api_cb($api_cb);
    
    return $self;
}

### PUBLIC INSTANCE METHOD ###
#
# Call specified Action's Method asynchronously
#

sub call_async { shift->async_request('call', @_) }

### PUBLIC INSTANCE METHOD ###
#
# Submit a form to specified Action's Method asynchronously
#

sub submit_async { shift->async_request('form', @_) }

### PUBLIC INSTANCE METHOD ###
#
# Upload a file using POST form. Same as submit()
#

*upload_async = *submit_async;

### PUBLIC INSTANCE METHOD ###
#
# Poll server for events asynchronously
#

sub poll_async { shift->async_request('poll', @_) }

### PUBLIC INSTANCE METHOD ###
#
# Run a specified request type asynchronously
#

sub async_request {
    my $self = shift;
    my $type = shift;
    
    my $tr_class    = $self->transaction_class;
    my $transaction = $tr_class->new(@_);
    
    #
    # We try to avoid action-at-a-distance here, so we will
    # call all the stuff that could die() up front,
    # to pass on the exception to the caller immediately
    # rather than blowing up later on.
    #
    eval { $self->_async_request($type, $transaction) };
    
    if ($@) { croak 'ARRAY' eq ref($@) ? $@->[0] : $@ };
    
    # Stay positive
    return 1;
}

### PUBLIC INSTANCE METHOD ###
#
# Return the name of the Transaction class
#

sub transaction_class { 'RPC::ExtDirect::Client::Async::Transaction' }

### PUBLIC INSTANCE METHOD ###
#
# Read-write accessor
#

RPC::ExtDirect::Util::Accessor->mk_accessor(
    simple => [qw/ api_ready api_cb request_queue /],
);

############## PRIVATE METHODS BELOW ##############

### PRIVATE INSTANCE METHOD ###
#
# Throw an exception using the condvar passed to the constructor,
# or just set an error so the next async request would die() with it
#

sub _throw {
    my ($self, $ex) = @_;
    
    my $cv = $self->cv;
    
    if ($cv) {
        $cv->croak($ex);
    }
    else {
        push @{$self->{exceptions}}, $ex;
    }
}

### PRIVATE INSTANCE METHOD ###
#
# Initialize API declaration
#

sub _init_api {
    my ($self) = @_;
    
    # We want to be truly asynchronous, so instead of
    # blocking on API retrieval, we create a request queue
    # and return immediately. If any call/form/poll requests happen
    # before we've got the API result back, we push them in the queue
    # and wait for the API to arrive, then re-run the requests.
    # After the API declaration has been retrieved, all subsequent
    # requests run without queuing.
    $self->request_queue([]);
    
    $self->_get_api(sub {
        my ($api_js) = @_;
        
        $self->_import_api($api_js);
        $self->api_ready(1);
        
        my $queue = $self->request_queue;
        delete $self->{request_queue};  # A bit quirky
        
        $_->() for @$queue;
        
        my $cv = $self->cv;
    
        $cv->end if $cv;
        
        $self->api_cb->($self) if $self->api_cb;
    });
}

### PRIVATE INSTANCE METHOD ###
#
# Receive API declaration from specified server,
# parse it and return Client::API object
#

sub _get_api {
    my ($self, $cb) = @_;

    my $cv     = $self->cv;
    my $uri    = $self->_get_uri('api');
    my $params = $self->{http_params};

    # Run additional checks before firing curried callback
    my $api_cb = sub {
        my ($content, $headers) = @_;
        
        $self->_throw("Can't download API declaration: $headers->{Status}\n")
            unless $headers->{Success} ne '200';

        $self->_throw("Empty API declaration\n")
            unless length $content;
        
        $cb->($content);
    };
    
    $cv->begin if $cv;

    AnyEvent::HTTP::http_request(
        GET => $uri,
        %$params,
        $api_cb,
    );
}

### PRIVATE INSTANCE METHOD ###
#
# Run asynchronous request(s) if the API is already available;
# queue for later if not
#

sub _run_request {
    my $self = shift;
    
    if ( $self->api_ready ) {
        $_->() for @_;
    }
    else {
        $self->_queue_request(@_);
    }
}

### PRIVATE INSTANCE METHOD ###
#
# Queue asynchronous request(s)
#

sub _queue_request {
    my $self = shift;
    
    my $queue = $self->{request_queue};
    
    push @$queue, @_;
}

### PRIVATE INSTANCE METHOD ###
#
# Make an HTTP request in asynchronous fashion
#

sub _async_request {
    my ($self, $type, $transaction) = @_;
    
    $self->_run_request(sub {
        my $prepare = "_prepare_${type}_request";
        my $method  = $type eq 'poll' ? 'GET' : 'POST';
    
        $transaction->start;
        
        my ($uri, $request_content, $http_params, $request_options)
            = eval { $self->$prepare($transaction) };
        
        $transaction->finish('ARRAY' eq ref $@ ? $@->[0] : $@, !1)
            if $@;
    
        my $request_headers = $request_options->{headers};
    
        # TODO Handle errors
        AnyEvent::HTTP::http_request(
            $method, $uri,
            headers => $request_headers,
            body    => $request_content,
            %$http_params,
            $self->_curry_response_cb($type, $transaction),
        )
    });
}

### PRIVATE INSTANCE METHOD ###
#
# Parse cookies if provided, creating Cookie header
#

sub _parse_cookies {
    my ($self, $to, $from) = @_;
    
    $self->SUPER::_parse_cookies($to, $from);
    
    # This results in Cookie header being a hashref,
    # but we need a string for AnyEvent::HTTP
    if ( $to->{headers} && (my $cookies = $to->{headers}->{Cookie}) ) {
        $to->{headers}->{Cookie} = join '; ', @$cookies;
    }
}

### PRIVATE INSTANCE METHOD ###
#
# Generate result handling callback
#

sub _curry_response_cb {
    my ($self, $type, $transaction) = @_;
    
    return sub {
        my ($data, $headers) = @_;
        
        my $status  = $headers->{Status};
        my $success = $status eq '200';
        
        my $handler  = "_handle_${type}_response";
        my $response = eval {
            $self->$handler({
                status  => $status,
                success => $success,
                content => $data,
            })
        } if $success;
        
        # We're only interested in the data, but anything goes
        my $result = 'ARRAY' eq ref($@)       ? $@->[0]
                   : $@                       ? $@
                   : !$success                ? $headers->{Reason}
                   : 'poll' eq $type          ? $response
                   : 'HASH' eq ref($response) ? $response->{result}
                   :                            $response
                   ;
        
        $transaction->finish($result, !$@ && $success);
    };
}

package
    RPC::ExtDirect::Client::Async::Transaction;

use Carp;

use base 'RPC::ExtDirect::Client::Transaction';

my @fields = qw/ cb cv actual_arg fields /;

sub new {
    my ($class, %params) = @_;
    
    croak "Callback subroutine is required"
        unless 'CODE' eq ref $params{cb};
    
    my %self_params = map { $_ => delete $params{$_} } @fields;
    
    my $self = $class->SUPER::new(%params);
    
    @$self{ keys %self_params } = values %self_params;
    
    return $self;
}

sub start {
    my ($self) = @_;
    
    my $cv = $self->cv;
    
    $cv->begin if $cv;
}

sub finish {
    my ($self, $result, $success) = @_;
    
    my $cb = $self->cb;
    my $cv = $self->cv;
    
    $cb->($result, $success) if $cb;
    $cv->end                 if $cv;
}

RPC::ExtDirect::Util::Accessor->mk_accessors(
    simple => [qw/ cb cv /],
);

1;

__END__

=pod

=head1 NAME

RPC::ExtDirect::Client - Ext.Direct client in Perl

=head1 SYNOPSIS

 use RPC::ExtDirect::Client;
 
 my $client = RPC::ExtDirect::Client->new(host => 'localhost');
 my $result = $client->call(
    action  => 'Action',
    method  => 'Method',
    arg     => [ 'foo', 'bar' ],
    cookies => { foo => 'bar' },
 );

=head1 DESCRIPTION

This module implements Ext.Direct client in pure Perl. Its main purpose
is to be used for testing server side Ext.Direct classes.

RPC::ExtDirect::Client uses HTTP::Tiny as transport.

=head1 METHODS

=over 4

=item new(%params)

Creates a new client instance. Constructor accepts the following arguments:

=over 8

=item api_path

URI for Ext.Direct API published by server. Default: '/api'.

=item router_path

URI for Ext.Direct remoting requests. Default: '/router'.

=item poll_path

URI for Ext.Direct events. Default: '/events'.

=item remoting_var

JavaScript variable name used to assign Ext.Direct remoting API object to.
Default: 'Ext.app.REMOTING_API'.

=item polling_var

JavaScript variable name used to assign Ext.Direct polling API object to.
Default: 'Ext.app.POLLING_API'.

=item cookies

Cookies to set when calling server side; can be either HTTP::Cookies object
or a hashref containing key-value pairs. Setting this in constructor will
pass the same cookies to all subsequent client calls.

=item %other

All other arguments are passed to HTTP::Tiny constructor. See L<HTTP::Tiny>
for more detail.

=back

=item get_api

Returns L<RPC::ExtDirect::Client::API> object with Ext.Direct API
declaration published by the server.

=item call(%params)

Calls Ext.Direct remoting method. Arguments are:

=over 8

=item action

Ext.Direct Action (class) name

=item method

Ext.Direct Method name to call

=item arg

Ext.Direct Method arguments; use arrayref for methods that accept ordered
parameters or hashref for named parameters.

=item cookies

Same as with constructor, but sets cookies for this particular call only.

=back

Returns either call Result or Exception.

=item submit

Submits a form request to formHandler method. Arguments should be:

=over 8

=item action

Ext.Direct Action (class) name

=item method

Ext.Direct Method name

=item arg

Method arguments; for formHandlers it should always be a hashref.

=item upload

Arrayref of file names to upload.

=item cookies

Same as with constructor, but sets cookies for this particular call only.

=back

Returns either call Result or Exception.

=item upload

Same as C<submit>.

=item poll

Polls server side for events, returns event data.

=over 8

=item cookies

Same as with constructor, but sets cookies for this particular call only.

=back

=back

=head1 DEPENDENCIES

RPC::ExtDirect::Client depends on the following modules:
L<HTTP::Tiny>, L<JSON>, and L<RPC::ExtDirect::Server> for testing.

=head1 SEE ALSO

For more information on using Ext.Direct with Perl, see L<RPC::ExtDirect>.
L<RPC::ExtDirect::Server> can be used to provide lightweight drop-in for
production environment to run Ext.Direct tests.

=head1 BUGS AND LIMITATIONS

There are no known bugs in this module. Use github tracker to report bugs
(the best way) or just drop me an e-mail. Patches are welcome.

=head1 AUTHOR

Alexander Tokarev E<lt>tokarev@cpan.orgE<gt>

=head1 ACKNOWLEDGEMENTS

I would like to thank IntelliSurvey, Inc for sponsoring my work
on this module.

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2012-2013 Alexander Tokarev.

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. See L<perlartistic>.

=cut



Makefile.PL
use 5.006000;
use strict;
use warnings;

use ExtUtils::MakeMaker;

# Add the `devtest` target to run regression and POD tests in one go
sub MY::postamble {
    return <<'END';
devtest :
	REGRESSION_TESTS=1 POD_TESTS=1 $(MAKE) test

END
}

# Override `disttest` so it would behave as `devtest`
sub MY::dist_test {
    return <<'END';
disttest : distdir
	cd $(DISTVNAME) && $(ABSPERLRUN) Makefile.PL
	cd $(DISTVNAME) && $(MAKE) $(PASTHRU)
	cd $(DISTVNAME) && $(MAKE) devtest $(PASTHRU)

END
}

my $MM_VERSION  = $ExtUtils::MakeMaker::VERSION;
my $github_repo = 'https://github.com/nohuhu/RPC-ExtDirect-Client';

my %CORE_REQ = (
    'RPC::ExtDirect' => '3.21',
    'HTTP::Tiny'     => '0.014',
);

my %TEST_REQ = (
    'Test::More'             => '0.82', # for explain()
    'CGI::ExtDirect'         => '3.20',
    'RPC::ExtDirect::Server' => '1.20',
);

WriteMakefile(
    NAME         => 'RPC::ExtDirect::Client',
    VERSION_FROM => 'lib/RPC/ExtDirect/Client.pm',
    ABSTRACT     => 'Ext.Direct client in Perl',
    AUTHOR       => 'Alex Tokarev <tokarev@cpan.org>',
    LICENSE      => 'perl',
    
    ($MM_VERSION >= 6.64
        ? (
            TEST_REQUIRES => \%TEST_REQ,
            PREREQ_PM     => \%CORE_REQ,
        )
        : (
            ($MM_VERSION >= 6.5503
                ? (
                    BUILD_REQUIRES => \%TEST_REQ,
                    PREREQ_PM      => \%CORE_REQ,
                )
                : (
                    PREREQ_PM => {
                        %TEST_REQ,
                        %CORE_REQ,
                    },
                )
            ),
        ),
    ),
    
    ($MM_VERSION >= 6.48 ? ( MIN_PERL_VERSION => 5.006000, ) : ()),

    ($MM_VERSION >= 6.46
        ? ( META_MERGE => {
                resources => {
                    bugtracker => "$github_repo/issues",
                    repository =>  $github_repo,
                },
            },
        )
        : ()
    ),
);

